internal/querynode/impl.go

```diff
@@ -90,6 +90,187 @@ func (node *QueryNode) GetStatisticsChannel(ctx context.Context) (*milvuspb.Stri
+			SegmentIDs:      req.SegmentIDs,
+			FromShardLeader: req.FromShardLeader,
+			Scope:           req.Scope,
+		}
+		runningGp.Go(func() error {
+			ret, err := node.getStatisticsWithDmlChannel(runningCtx, req, ch)
+			mu.Lock()
+			defer mu.Unlock()
+			if err != nil {
+				failRet.Status.Reason = err.Error()
+				failRet.Status.ErrorCode = commonpb.ErrorCode_UnexpectedError
+				return err
+			}
+			if ret.GetStatus().GetErrorCode() != commonpb.ErrorCode_Success {
+				failRet.Status.Reason = ret.Status.Reason
+				failRet.Status.ErrorCode = ret.Status.ErrorCode
+				return fmt.Errorf("%s", ret.Status.Reason)
+			}
+			toReduceResults = append(toReduceResults, ret)
+			return nil
+		})
+	}
+	if err := runningGp.Wait(); err != nil {
+		return failRet, nil
+	}
+	ret, err := reduceStatisticResponse(toReduceResults)
+	if err != nil {
+		failRet.Status.ErrorCode = commonpb.ErrorCode_UnexpectedError
+		failRet.Status.Reason = err.Error()
+		return failRet, nil
+	}
+	return ret, nil
+}
+
+func (node *QueryNode) getStatisticsWithDmlChannel(ctx context.Context, req *queryPb.GetStatisticsRequest, dmlChannel string) (*internalpb.GetStatisticsResponse, error) {
+	failRet := &internalpb.GetStatisticsResponse{
+		Status: &commonpb.Status{
+			ErrorCode: commonpb.ErrorCode_UnexpectedError,
+		},
+	}
+
+	if !node.isHealthy() {
+		failRet.Status.Reason = msgQueryNodeIsUnhealthy(Params.QueryNodeCfg.GetNodeID())
+		return failRet, nil
+	}
+
+	msgID := req.GetReq().GetBase().GetMsgID()
+	log.Debug("received GetStatisticRequest",
+		zap.Int64("msgID", msgID),
+		zap.Bool("fromShardLeader", req.GetFromShardLeader()),
+		zap.String("vChannel", dmlChannel),
+		zap.Int64s("segmentIDs", req.GetSegmentIDs()),
+		zap.Uint64("guaranteeTimestamp", req.GetReq().GetGuaranteeTimestamp()),
+		zap.Uint64("timeTravel", req.GetReq().GetTravelTimestamp()))
+
+	if node.queryShardService == nil {
+		failRet.Status.Reason = "queryShardService is nil"
+		return failRet, nil
+	}
+
+	qs, err := node.queryShardService.getQueryShard(dmlChannel)
+	if err != nil {
+		log.Warn("get statistics failed, failed to get query shard",
+			zap.Int64("msgID", msgID),
+			zap.String("dml channel", dmlChannel),
+			zap.Error(err))
+		failRet.Status.ErrorCode = commonpb.ErrorCode_NotShardLeader
+		failRet.Status.Reason = err.Error()
+		return failRet, nil
+	}
+
+	log.Debug("start do statistics",
+		zap.Int64("msgID", msgID),
+		zap.Bool("fromShardLeader", req.GetFromShardLeader()),
+		zap.String("vChannel", dmlChannel),
+		zap.Int64s("segmentIDs", req.GetSegmentIDs()))
+	tr := timerecord.NewTimeRecorder("")
+
+	waitCanDo := func() error {
+		ch := node.tSafeReplica.WatchChannel(dmlChannel)
+		defer node.tSafeReplica.UnwatchChannel(dmlChannel)
+		for {
+			serviceTime, err := qs.getServiceableTime(dmlChannel)
+			if err != nil {
+				return err
+			}
+			guaranteeTs := req.GetReq().GetGuaranteeTimestamp()
+			if guaranteeTs <= serviceTime {
+				return nil
+			}
+			<-ch
```

Comment:
need to select ctx, if there is cancel outside this, you can exit this loop in advance.

Reply:
There is  one check on `querynode/task_statistic.go`:96
```go
func (s *statistics) Execute(ctx context.Context) error {
	if err := s.waitCanDoFunc(); err != nil {
		return err
	}
	// check ctx timeout
	if !funcutil.CheckCtxValid(s.ctx) {
		return errors.New("get statistics context timeout")
	}
	if s.scope == querypb.DataScope_Streaming {
		return s.statisticOnStreaming()
	} else if s.scope == querypb.DataScope_Historical {
		return s.statisticOnHistorical()
	}
	return fmt.Errorf("statistics do not implement do statistic on all data scope")
}
```